From 0f3e37997597bd1a828762babc9eaffacb3cf92b Mon Sep 17 00:00:00 2001
From: Ben Greear <greearb@candelatech.com>
Date: Tue, 29 Jan 2013 22:48:24 -0800
Subject: [PATCH 04/37] nfs: Use request destination addr as callback source
 addr.

This helps make sure that responses use the correct
source address on multi-homed machines.

Signed-off-by: Ben Greear <greearb@candelatech.com>
---
 fs/nfs/callback.h     | 3 +++
 fs/nfs/callback_xdr.c | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/fs/nfs/callback.h b/fs/nfs/callback.h
index a20a0bc..24862c5 100644
--- a/fs/nfs/callback.h
+++ b/fs/nfs/callback.h
@@ -63,6 +63,7 @@ struct cb_compound_hdr_res {
 };
 
 struct cb_getattrargs {
+	struct sockaddr *srcaddr;
 	struct nfs_fh fh;
 	uint32_t bitmap[2];
 };
@@ -77,6 +78,7 @@ struct cb_getattrres {
 };
 
 struct cb_recallargs {
+	struct sockaddr *srcaddr;
 	struct nfs_fh fh;
 	nfs4_stateid stateid;
 	uint32_t truncate;
@@ -97,6 +99,7 @@ struct referring_call_list {
 
 struct cb_sequenceargs {
 	struct sockaddr			*csa_addr;
+	struct sockaddr                 *csa_daddr;
 	struct nfs4_sessionid		csa_sessionid;
 	uint32_t			csa_sequenceid;
 	uint32_t			csa_slotid;
diff --git a/fs/nfs/callback_xdr.c b/fs/nfs/callback_xdr.c
index 57de914..3eae3e1 100644
--- a/fs/nfs/callback_xdr.c
+++ b/fs/nfs/callback_xdr.c
@@ -189,6 +189,7 @@ static __be32 decode_getattr_args(struct svc_rqst *rqstp,
 	status = decode_fh(xdr, &args->fh);
 	if (unlikely(status != 0))
 		return status;
+	args->srcaddr = svc_daddr(rqstp);
 	return decode_bitmap(xdr, args->bitmap);
 }
 
@@ -199,6 +200,7 @@ static __be32 decode_recall_args(struct svc_rqst *rqstp,
 	__be32 *p;
 	__be32 status;
 
+	args->srcaddr = svc_daddr(rqstp);
 	status = decode_delegation_stateid(xdr, &args->stateid);
 	if (unlikely(status != 0))
 		return status;
@@ -420,6 +422,7 @@ static __be32 decode_cb_sequence_args(struct svc_rqst *rqstp,
 		return htonl(NFS4ERR_RESOURCE);
 
 	args->csa_addr = svc_addr(rqstp);
+	args->csa_daddr = svc_daddr(rqstp);
 	args->csa_sequenceid = ntohl(*p++);
 	args->csa_slotid = ntohl(*p++);
 	args->csa_highestslotid = ntohl(*p++);
-- 
2.4.11

